### PR TITLE
Generalize ldap for more flexible use

### PR DESCRIPTION
**Pull request by Skywalker-11**
More generalized version for ldap authentication. Should make use cases like #549 possible by configuration (no code change needed).

### CODE REVIEW

**Comment by remkolodder** [comment]
I think this might be sufficient, but the current proposed usage is only in conf.php, while I think we should modify functions.php as well to actually use it?

I also spotted a little typo, will comment on that.

File: mailscanner/functions.php [hunk 1]
```diff
@@ -2677,7 +2675,7 @@ function ldap_authenticate($user, $password)
                     return null;
                 }
 
-                $bindResult = @ldap_bind($ds, $user, $password);
+                $bindResult = @ldap_bind($ds, $userlogin, $password);
```
Comment (remkolodder) [inline mailscanner/functions.php]: $userlogin does not exist, this should remain $user .

File: mailscanner/functions.php [hunk 2]
```diff
@@ -2664,11 +2656,17 @@ function ldap_authenticate($user, $password)
...
+                      !isset($result[0][LDAP_USERNAME_FIELD][0])) {
+                    @trigger_error(__('ldapno03') . ' "' . LDAP_USERNAME_FIELD . '" ' . __('ldapresults03'));
+                    return null;
+                }
+
+                if (LDAP_USERNAME_FIELD === 'cn') {
                     // build DN for user, when LDAP server is not an AD
                     $user = 'cn=' . $result[0]['cn'][0] . ',' . LDAP_DN;
+                } else {
+                    $user = $result[0][LDAP_USERNAME_FIELD][0];
```
Comment (remkolodder) [inline mailscanner/functions.php]: This way, the $result will just be the field filled in, but a search needs the
complete DN.

for me this is needed;

> $user = LDAP_USERNAME_FIELD . '=' . $result[0][LDAP_USERNAME_FIELD][0] . ',' . LDAP_DN;

File: mailscanner/functions.php (hunk 1 above)
Comment (Skywalker-11) [inline mailscanner/functions.php]: done

File: mailscanner/functions.php (hunk 2 above)
Comment (Skywalker-11) [inline mailscanner/functions.php]: Introduced prefix and suffix because to enable to do this.

**Comment by endelwar** [comment]
Very good!
Some config examples should be provided in docs for major user case (Active Directory or openldap standard installations)

**Pull request merged**